### PR TITLE
[#67] Do not decode binary response body

### PR DESCRIPTION
**Pull request by robertoaloi**
Emulating the approach in #62 , this PR skips decoding of the response body for those methods where the body is a binary.
Since my Elixir knowledge is quite limited, please advise on the change for the Elixir generated code (currently not included in this PR).

### CODE REVIEW

File: priv/rest.erl.eex [hunk 1]
```diff
@@ -154,14 +162,17 @@ handle_response({ok, StatusCode, ResponseHeaders, Client}, SuccessStatusCode)
                         StatusCode =:= SuccessStatusCode ->
             {ok, #{}, {StatusCode, ResponseHeaders, Client}};
         {ok, Body} ->
-            Result = <%= context.decode %>,
+            Result = case DecodeBody of
+                       true -> <%= context.decode %>;
+                       false -> #{<<"Body">> => Body}
```
Comment (robertoaloi) [inline priv/rest.erl.eex:167]: Apparently, the code expects this to be a map. The generated Erlang code for an API looks like this:

```
    case request(Client, Method, Path, Query_, Headers, Input, Options, SuccessStatusCode) of
      {ok, Body0, {_, ResponseHeaders, _} = Response} ->
        ResponseHeadersParams =
          [
            {<<"x-amz-server-side-encryption-bucket-key-enabled">>, <<"BucketKeyEnabled">>},
            [...]
            {<<"x-amz-version-id">>, <<"VersionId">>}
          ],
        FoldFun = fun({Name_, Key_}, Acc_) ->
                      case lists:keyfind(Name_, 1, ResponseHeaders) of
                        false -> Acc_;
                        {_, Value_} -> Acc_#{Key_ => Value_}
                      end
                  end,
        Body = lists:foldl(FoldFun, Body0, ResponseHeadersParams),
        {ok, Body, Response};
      Result ->
        Result
    end.
```

So, it looks like we merge the "body" with the "response headers". This does not look too nice, also not sure `<<"Body">>` is the key we want. But I guess this should be addressed separately from this PR. Any insights @jfacorro ?

**Review by philss — COMMENTED** [review]
LGTM! :+1: 

Only a minor suggestion.
I think the Elixir version requires the update of the `AWS.Request` module:
- https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/request.ex#L44
- https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/request.ex#L104

File: lib/aws_codegen/shapes.ex [hunk 2]
```diff
@@ -5,7 +5,11 @@ defmodule AWS.CodeGen.Shapes do
     get_in(operation_spec, ["input", "shape"])
   end
 
-  def send_body_as_binary?(shapes, input_shape) do
+  def get_output_shape(operation_spec) do
+    get_in(operation_spec, ["output", "shape"])
+  end
+
+  def body_as_binary?(shapes, input_shape) do
```
[the remark this answers is not in the document]
Comment (robertoaloi) [reply lib/aws_codegen/shapes.ex]: Nice catch, renamed.

**Comment by robertoaloi** [comment]
> I think the Elixir version requires the update of the `AWS.Request` module:
> 
> * https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/request.ex#L44
> * https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/request.ex#L104

Let me know if you would like me to also check the Elixir parts or if you prefer tackling that yourself. I'm new to Elixir, so that's the slowest part for me...

**Comment by robertoaloi** [comment]
> Let me know if you would like me to also check the Elixir parts or if you prefer tackling that yourself. I'm new to Elixir, so that's the slowest part for me...

Created #71 for the Elixir part, not to be blocked on the Erlang side. I will try to address it, but if you want to jump in you're welcome!

**Pull request merged**